### ISSUE TITLE
<script type="py" worker>...</script> doesn't run in a worker

### ISSUE THREAD

**Issue by antocuni** [pyscript/pyscript#1664]
### Checklist

- [X] I added a descriptive title
- [X] I searched for other issues and couldn't find a solution or duplication
- [X] I already searched in Google and didn't find any good information or help

### What happened?

I tried to run the following snippet:
```html
<script type="py" worker>
    import js
    IN_MAIN = hasattr(js, 'document')
    print('in main thread?', IN_MAIN)
</script>
```

I expected that it would run the python code in a worker, but it doesn't seem so. What I see is that the main thread blocks while loading pyodide, and it prints `in main thread? True`.

I tried to move the code in a separate file `foo.py`, and the following works as expected:
```html
<script type="py" worker="foo.py"></script>
```

### What browsers are you seeing the problem on? (if applicable)

Chrome

**Comment by WebReflection** [comment]
This is by design and also documented: https://pyscript.github.io/polyscript/#script-attributes

> Bootstraps an interpreter only within a worker, allowing config and version attributes too, also attaching an xworker property/field directly to the script tag on the main page. Please note the interpreter will not be available on the main thread when this attribute is used.

I find your expectations a bit of a stretch in terms of separation of concerns, you have code in main that never executes in main ... there's nothing like that in the *JS* world and it would be *very* misleading for people with current "classic" py-script beleving that just adding `worker` attribute everything will magically work the same as expected.

The reality is that `import js` and the usage of `js.document` among every other possibly missing `js.XXX` utilities that don't exist in workers will break ... so while I have thought about your use case or expectations, I am myself not sure that is the right thing to do.

**Comment by WebReflection** [comment]
P.S. if we want to fix this, the right place is polyscript, not here.

**Comment by antocuni** [comment]
> This is by design and also documented: https://pyscript.github.io/polyscript/#script-attributes

uhm, I remember that @fpliger showed me an example of using `worker` as a standalone attribute. So, it's entirely possible that I misunderstood what he showed me, or it's possible that he also expected `<script type="py" worker>` to run in a worker.

We can discuss whether this should be allowed or not (see below), but in case we decide that it's not allowed, it should emit a clear error, instead of silently ignoring the attribute, IMHO.


> I find your expectations a bit of a stretch in terms of separation of concerns, you have code in main that never executes in main ... there's nothing like that in the _JS_ world and it would be _very_ misleading for people with current "classic" py-script beleving that just adding `worker` attribute everything will magically work the same as expected.

I have been advocating for ages that we should run python code in a worker *by default* :).
I agree with you that it might be unexpected for people coming from the JS world. But at the same time, python-in-the-main-thread also behaves very unexpectedly for all the people coming from the Python world (because it misses all the sync APIs).

It's very likely that we need to make a choice and decide what is the category of users that we want to surprise the least "by default".

**Comment by WebReflection** [comment]
in recent *polyscript* discussions (early users, tons of valuable feedback) worker by default is not desired at all ... having `worker` attribute that forward the code (I think it should "URL-Blob" it as secure content instead) might make sense and provide best expectations or least surprise ... the fact @fpliger also showed you that, believing anything was different, tells a story ... 2 PyScript folks expectations burned (but it was documented though, you gotta give me that).

I am OK making `worker` that magic, but then again, let's move the discussion where it belongs: [polyscript](https://github.com/pyscript/polyscript)

**Comment by antocuni** [comment]
I 50% agree with you :).

As a polyscript user, I'd not expect the code to run in a worker by default.
But on the other hand, I suspect that the typical polyscript user is very different than the typical pyscript user: the former is probably more experienced and knows a lot about the web, while many pyscript users are total beginners and don't know anything about the web.

For *them*, it might make sense to have python-in-the-worker by default, because it provides a better/easier experience.
So, from this point of view, it might be more appropriate to have the discussion here

I also admit that having such a different behavior between polyscript and pyscript is non ideal, so if we find a better solution I'd be very happy to hear that.

**Comment by WebReflection** [comment]
it's not that is not ideal, *pyscript* is simply a *polyscript* custom type ... whatever we decide here, requires changes there because everything about the DOM is handled in there, including attributes ... so if it has to inform custom types there was a `worker` one and the resulting url is "that one", either it does it for all interpreters or we need to fork *polyscript* for a pyscript only usage, which I don't recommend at all, as that smells from an integration point of view to me.

**Comment by WebReflection** [comment]
last thoughts: before going workers by default I think we should release ... if we go workers by default we can say goodbye to feature parity to start with, imho ... the worker only story also has things that are different for both plugins and what's not, because worker hooks are only strings. I think we should focus on creating the best we can based on what worked so far without exaggerating with "breaking the world". Remember that worker DOM access is also inevitably slower due Atomics and postMessage dance and there's no fix or hack we can do to to bypass that dance.

**Comment by antocuni** [comment]
> last thoughts: before going workers by default I think we should release 

yes, 100%.
workers by default should be something which happens in the future. In my ideal world, what I would like is:

1. for the upcoming release, `<py-script>` runs in main and `<py-script worker>` works
2. for a future release, `<py-script>` runs in workers by default and you use something `<py-script main>` to explicitly opt-in for the main thread

But note that these are only my personal thoughts, I don't know what is the position of the rest of the team

**Comment by fpliger** [comment]
> uhm, I remember that @fpliger showed me an example of using worker as a standalone attribute. So, it's entirely possible that I misunderstood what he showed me, or it's possible that he also expected <script type="py" worker> to run in a worker.

I recall that conversation and, yes, I showed you something like `<py-script worker="filename.py">` or `<script type="py" worker="foo.py"></script>` and iirc, I've mentioned something like "we don't have a way to run inline code in a worker as easily. I think it'd be cool if we allowed to just specify the `worker` attribute for that case too".

I really didn't mean to imply that was an issue or a release blocker. It definitely is not.. just part of the process of us sharing ideas of things we can improve in the future.  Sorry if that wasn't clear.

> > last thoughts: before going workers by default I think we should release

> yes, 100%.

double 100% from me too  :)
